kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/PulsarStorageProducerIdManagerImpl.java

```diff
@@ -0,0 +1,197 @@
+                throw new CompletionException(b);
+            }
+        });
+    }
+
+    private synchronized void endReadLoop(CompletableFuture<?> handle) {
+        if (handle == currentReadHandle) {
+            currentReadHandle = null;
+        }
+    }
+
+    @Override
+    public synchronized CompletableFuture<Long> generateProducerId() {
+        Long booked = availableIdsLocally.poll();
+        if (booked != null) {
+            log.debug("Returning pre-allocated id {} for {}", booked, topic);
+            return CompletableFuture.completedFuture(booked);
+        }
+        log.debug("Allocating new block of ids for {}", topic);
+        // we could get rid of the Exclusive Producer if we had Message.getIndex()
+        // introduced in 2.9.0 https://github.com/apache/pulsar/pull/11553
+        CompletableFuture<Producer<byte[]>> producerHandle = pulsarClient.newProducer()
+                .enableBatching(false)
+                .topic(topic)
+                .accessMode(ProducerAccessMode.WaitForExclusive)
+                .blockIfQueueFull(true)
+                .createAsync();
+        return producerHandle.thenCompose(opProducer -> {
+            // nobody can write now to the topic
+            // wait for local cache to be up-to-date
+            CompletableFuture<Long> dummy = ensureLatestData(true)
+                    .thenCompose((___) -> {
+                        long start = nextId.get();
+                        List<Long> block = generateBlock(start + blockSize, blockSize);
+                        final long nextAvailableId = start + blockSize;
+                        // write to Pulsar
+                        byte[] serialized = BigInteger.valueOf(nextAvailableId).toByteArray();
+                        CompletableFuture<Long>  res =  opProducer
+                                .newMessage()
+                                .key("") // always the same key, this way we can rely on compaction
+                                .value(serialized)
+                                .sendAsync()
+                                .thenApply((msgId) -> {
+                                    log.debug("{} written {} as {}", this, nextAvailableId, msgId);
+                                    nextId.set(nextAvailableId);
+                                    availableIdsLocally.addAll(block);
+                                    long result = availableIdsLocally.remove();
+                                    log.debug("Returning allocated id {} for {}, new range: {}-{}",
+                                            result, topic, start, nextAvailableId);
+                                    return result;
+                                });
+                        return res;
+                    });
+            // ensure that we release the exclusive producer in any case
+            dummy.whenComplete((___, err) -> {
+                opProducer.closeAsync();
+            });
+            return dummy;
```

Comment:
Maybe we can return the future returned by `whenComplete` like:

```java
            return dummy.whenComplete((___, err) -> {
                opProducer.closeAsync();
            });
```

In addition, it's better to print the error message if `closeAsync` failed.

Reply:
good idea